compiler/rustc_feature/src/builtin_attrs.rs

```diff
@@ -458,6 +458,8 @@ pub const BUILTIN_ATTRIBUTES: &[BuiltinAttribute] = &[
     ),
     // Enumerates "identity-like" conversion methods to suggest on type mismatch.
     rustc_attr!(rustc_conversion_suggestion, AssumedUsed, template!(Word), INTERNAL_UNSTABLE),
+    // Causes field reads in the annotated item to be ignored during dead code analysis.
+    ungated!(rustc_trivial_field_reads, AssumedUsed, template!(Word)),
```

Comment:
wait, why is this `ungated!`? that doesn't seem right

Reply:
doesn't that mean anybody can use it? that's not what we want :)

Reply:
I initially thought so, too, but remember that this is generated in user code as part of the derive macro expansion, so anybody _has_ to be able to use it in order to use the derive macro (in my initial implementation, I gated the attribute and received lots of errors about uses of unstable/internal features in the expanded user code). Or is there a more elegant way of resolving this problem?

Reply:
Well, we have mechanisms for derives and things to use unstable code, but an alternative would be to put the attribute on the **trait**, rather than the impl. We would then detect an auto-generated impl for a trait with that attribute. (Plausibly this could apply to hand-written code, too...)

Reply:
I think I would prefer the attribute on the trait, I think it'd be more visible, and it's more obvious how to extend to other traits should we decide to do so in the future.

Reply:
Thoughts, @FabianWolff ?

Reply:
@nikomatsakis Done. Sorry for taking so long to implement this.

I have decided not to extend this to manual implementations, though, i.e. my latest changes still only apply to automatically derived `impl`s. I think this would otherwise become too confusing/annoying; for instance, when ignoring all `Clone` implementations, including, say, this one:
https://github.com/rust-lang/rust/blob/69c4aa2901ffadf69deaf91b2f90604bcbc2eb36/library/core/src/ffi.rs#L364-L371

the `va_copy()` function would be reported as dead code, but we can't actually delete it, because it _is_ used. The same situation can arise for field reads if someone chooses to manually implement `Clone`/`Debug` for whatever reason.

Reply:
in our codebase, we've recently discovered that some of our `proc_macro_derive`s have been hiding dead code. The derive uses fields to create an implementation for a different structure but fields must be accessed in order for that to have any use.

Am I correct in understanding that after `rustc_trivial_field_reads` has been gated, there was no trait or a different way to mark our derive to be ignored during dead code analysis? I only see something similar in shepmaster's [comment](https://github.com/rust-lang/rust/pull/85200#issuecomment-1010371394) lower, and that is never addressed. Also, I'm a bit confused as to where to ask about this problem, if I should open an issue somewhere, or if there is some workaround. 

Ironically, everything I was able to google up was talking about how to silence the warning, not the other way around.

Reply:
Update regarding my question, we've managed to redo derive a bit so that it does not read fields and restrict visibility to  get dead_code warnings back